karabo_data/reader.py

```diff
@@ -129,70 +175,134 @@ def _parse_data_src(source):
+                    'source': source,
+                    'timestamp': ts,
+                    'timestamp.tid': train_id,
+                    'timestamp.sec': sec,
+                    'frac': frac,
+                }
+            return train_id, train_data
+
         for source in self.sources:
-            h5_device = source.split('/', 1)[1]
-            index = self.index[h5_device]
+            h5_source = source.split('/', 1)[1]
+            index = self.index[h5_source]
             table = self.file[source]
 
             # Which parts of the data to get for this train:
-            first = int(index['first'][train_index])
-            if 'last' in index:
-                # Older (?) format: status (0/1), first, last
-                last = int(index['last'][train_index])
-                status = index['status'][train_index]
-            else:
-                # Newer (?) format: first, count
-                count = int(index['count'][train_index])
-                last = first + count - 1
-                status = count > 0
+            first, count = self._read_index(h5_source, train_index)
 
             category, device, path_base = self._parse_data_src(source)
 
-            if only_this and device not in only_this:
+            if not count:
                 continue
 
-            if device not in train_data:
-                train_data[device] = {}
             data = train_data[device]
 
-            if status:
-                def append_data(key, value):
-                    if isinstance(value, h5py.Dataset):
-                        path = '.'.join(filter(None,
-                                        (path_base,) + tuple(key.split('/'))))
-                        if (only_this and only_this[device] and
-                                path not in only_this[device]):
-                            return
+            def append_data(key, value):
+                if isinstance(value, h5py.Dataset):
+                    path = '.'.join(filter(None,
+                                    (path_base,) + tuple(key.split('/'))))
+
+                    if count == 1:
+                        data[path] = value[first]
+                    else:
+                        data[path] = value[first:first + count, ]
 
-                        if first == last:
-                            data[path] = value[first]
-                        else:
-                            data[path] = value[first:last+1, ]
+            table.visititems(append_data)
 
-                table.visititems(append_data)
 
-            sec, frac = str(time()).split('.')
-            timestamp = {'tid': int(self.train_ids[train_index]),
-                         'sec': int(sec), 'frac': int(frac)}
-            data.update({'metadata': {'source': device, 'timestamp': timestamp}})
+            train_data[device]['metadata'] = {
+                'source': source,
+                'timestamp': ts,
+                'timestamp.tid': int(self.train_ids[train_index]),
```

Comment:
better `train_id` instead of `int(self.train_ids[train_index])` since it is used everywhere else. Also why the casting? Does it return non-`int` type? if it is the case then I would do it line `215 train_id = int(self.train_ids[train_index])`